shell/platform/linux/fl_view.cc

```diff
@@ -241,6 +242,41 @@ static gboolean fl_view_button_release_event(GtkWidget* widget,
   return fl_view_send_pointer_button_event(self, event);
 }
 
+// Implements GtkWidget::scroll_event.
+static gboolean fl_view_scroll_event(GtkWidget* widget, GdkEventScroll* event) {
+  FlView* self = FL_VIEW(widget);
+
+  // TODO(robert-ancell): Update to use GtkEventControllerScroll when we can
+  // depend on GTK 3.24.
+
+  gdouble scroll_delta_x = 0.0, scroll_delta_y = 0.0;
+  if (event->direction == GDK_SCROLL_SMOOTH) {
+    scroll_delta_x = event->delta_x;
+    scroll_delta_y = event->delta_y;
+  } else if (event->direction == GDK_SCROLL_UP) {
+    scroll_delta_y = -1;
+  } else if (event->direction == GDK_SCROLL_DOWN) {
+    scroll_delta_y = 1;
+  } else if (event->direction == GDK_SCROLL_LEFT) {
+    scroll_delta_x = -1;
+  } else if (event->direction == GDK_SCROLL_RIGHT) {
+    scroll_delta_x = 1;
+  }
+
+  // TODO: See if this can be queried from the OS; this value is chosen
+  // arbitrarily to get something that feels reasonable.
+  const int kScrollOffsetMultiplier = 20;
```

Comment:
Should this be the same for smooth and non-smooth?

Reply:
For my mouse I'm getting magnitudes of 1.0 for both smooth and non-smooth events. Existing GTK code that uses scroll events uses the same step size for both smooth and non-smooth.

The GTK scrolling code seems very weird, you get both smooth and non-smooth events (though I think drivers technically may not send smooth events). This means you get two events for every scroll, which seems wrong. And the first smooth event is (0,0), which means the first scroll is half the distance of the following ones. I've copied the same method that the GTK widgets use for scrolls, so if it's wrong it's something that probably needs to be fixed in GTK.